src/lib/fetch.ts

```diff
@@ -452,6 +466,112 @@ function sendZkappQuery(json: string) {
+`;
+};
+
+type EventActionFilterOptions = {
+  to?: UInt32;
+  from?: UInt32;
+};
+
+/**
+Asynchronously fetches event data for an account from the Mina Archive Node GraphQL API.
+@async
+@param accountInfo - The account information object.
+@param accountInfo.publicKey - The account public key.
+@param [accountInfo.tokenId] - The optional token ID for the account.
+@param [graphqlEndpoint=archiveGraphqlEndpoint] - The GraphQL endpoint to query. Defaults to the Archive Node GraphQL API.
+@param [filterOptions={}] - The optional filter options object.
+@returns A promise that resolves to an array of objects containing event data and block height for the account.
+@throws If the GraphQL request fails or the response is invalid.
+@example
+const accountInfo = { publicKey: 'B62qiwmXrWn7Cok5VhhB3KvCwyZ7NHHstFGbiU5n7m8s2RqqNW1p1wF' };
+const events = await fetchEvents(accountInfo);
+console.log(events);
+*/
+async function fetchEvents(
+  accountInfo: { publicKey: string; tokenId?: string },
+  graphqlEndpoint = archiveGraphqlEndpoint,
+  filterOptions: EventActionFilterOptions = {}
+): Promise<any> {
+  if (!graphqlEndpoint)
+    throw new Error(
+      'fetchEvents: Specified GraphQL endpoint is undefined. Please specify a valid endpoint.'
+    );
+  const { publicKey, tokenId } = accountInfo;
+  let [response, error] = await makeGraphqlRequest(
+    getEventsQuery(
+      publicKey,
+      tokenId ?? TokenId.toBase58(TokenId.default),
+      filterOptions
+    ),
+    graphqlEndpoint
+  );
+  if (error) throw Error(error.statusText);
+  let fetchedEvents = response?.data.events;
+  if (fetchedEvents === undefined) {
+    throw Error(
+      `Failed to fetch events data. Account: ${publicKey} Token: ${tokenId}`
+    );
+  }
+
+  // TODO: This is a temporary fix. We should be able to fetch the event/action data from any block at the best tip.
+  // Once https://github.com/o1-labs/Archive-Node-API/issues/7 is resolved, we can remove this.
+  // If we have multiple blocks returned at the best tip (e.g. distanceFromMaxBlockHeight === 0),
+  // then filter out the blocks at the best tip. This is because we cannot guarantee that every block
+  // at the best tip will have the correct event data or guarantee that the specific block data will not
+  // fork in anyway. If this happens, we delay fetching event data until another block has been added to the network.
+  let numberOfBestTipBlocks = 0;
+  for (let i = 0; i < fetchedEvents.length; i++) {
+    if (fetchedEvents[i].blockInfo.distanceFromMaxBlockHeight === 0) {
+      numberOfBestTipBlocks++;
+    }
+    if (numberOfBestTipBlocks > 1) {
+      fetchedEvents = fetchedEvents.filter((event: any) => {
+        return event.blockInfo.distanceFromMaxBlockHeight !== 0;
+      });
+      break;
+    }
+  }
+
+  return fetchedEvents.map((event: any) => {
+    const events = event.eventData.map(
+      (eventData: { index: string; data: string[] }) => {
+        return [eventData.index].concat(eventData.data);
```

Comment:
We do this because `zkapp.fetchEvents` expects all the event data (the index of the event object and the data being emitted) in a single array. The GraphQL schema differentiates these values when you query, so I have added some code here to join them together from the GraphQL query in a format that the internal API expects, without changing too much.

Reply:
Actually, this was wrong :sweat_smile: Fixed to use `[eventData.index, ...eventData.data]` syntax

Reply:
Hmm, that's a good point. We can filter out the index type when we fetch from the API if there is only 1 type, that way only the calling function of the API handles that and the rest of the codebase doesn't have to change.